src/librustc_codegen_ironox/x86_asm_printer.rs

```diff
@@ -0,0 +1,79 @@
+use super::ModuleIronOx;
+
+use context::CodegenCx;
+use value::{Value, Instruction};
+
+use rustc::mir::mono::Stats;
+use rustc::util::nodemap::FxHashMap;
+use rustc_codegen_ssa::traits::MiscMethods;
+use std::cell::{Cell, RefCell};
+
+/// A convenience macro for appending instructions to a string.
+macro_rules! asm {
+    (
+        $m:ident,
+        $(
+            $fmt:expr; [ $($args:expr),* ]
+        ),*
+    ) => {
+        let mut asm_lines = vec![$(format!($fmt, $($args),*)),*];
+        asm_lines = asm_lines.iter().map(|x| format!("\t{}\n", x)).collect();
+        let asm_lines = asm_lines.join("");
+        $m.push_str(&asm_lines);
+    };
+    ($m:ident, $new_asm:expr) => {
+        $m.push_str($new_asm);
+    }
+}
+
+/// A convenience macro for appending a label to a string.
+macro_rules! label {
+    ($m:ident, $lbl:expr) => {
+        $m.push_str(&format!("{}:\n", $lbl));
+    }
+}
+
+/// The result of evaluating an `Instruction`.
+#[derive(Debug)]
+pub struct InstrAsm {
+    /// The sequence of assembly instructions the instruction is compiled to.
+    asm: String,
+    /// The register/memory address that contains the result of the instruction.
+    result: String,
+}
+
+/// A module that contains the context for generating machine instructions from
+/// a `ModuleIronOx`.
+pub struct ModuleAsm<'ll, 'tcx> {
+    /// The codegen context, which also contains the `ModuleIronOx` to be compiled.
+    cx: CodegenCx<'ll, 'tcx>,
+    /// A mapping from high-level instructions to the assembly they compile to.
+    /// If the same instruction is used in two different expressions, it is
+    /// often enough to retrieve its result from this mapping (it does not always
+    /// have to be recompiled).
+    compiled_insts: RefCell<FxHashMap<Value, InstrAsm>>,
+    /// A mapping from function parameters to their location on the stack.
+    /// Currently, parameters are always pushed to and retrieved from the stack.
+    compiled_params: RefCell<FxHashMap<Value, InstrAsm>>,
```

Comment:
How is the location on the stack expressed used asm?

Reply:
It's not ideal, but it's a string (for example, `0x10(%rbp)`).